server/src/main/java/io/spine/server/entity/Migration.java

```diff
@@ -0,0 +1,277 @@
+ *
+ * <p>To create a user-defined {@code Migration} in real life scenarios, consider inheriting from
+ * {@link io.spine.server.projection.ProjectionMigration} and
+ * {@link io.spine.server.procman.ProcessManagerMigration} types.
+ *
+ * @param <I>
+ *         the entity ID type
+ * @param <E>
+ *         the entity type
+ * @param <S>
+ *         the entity state type
+ */
+@Experimental
+public abstract class Migration<I, E extends TransactionalEntity<I, S, ?>, S extends EntityState>
+        implements Function<S, S>, Logging {
+
+    /**
+     * The currently performed migration operation.
+     */
+    private @Nullable Operation<I, S, E> currentOperation;
+
+    /**
+     * Applies the migration to a given entity, starting a new migration
+     * {@linkplain Operation operation}.
+     */
+    final void applyTo(E entity, RecordBasedRepository<I, E, S> repository) {
+        currentOperation = new Operation<>(entity, repository);
+
+        Transaction<I, E, S, ?> tx = txWithLifecycleMonitor();
+        S oldState = entity.state();
+        S newState = apply(oldState);
+        currentOperation().updateState(newState);
+        currentOperation().updateLifecycle();
+        tx.commit();
+    }
+
+    /**
+     * Marks the entity under migration as {@linkplain Entity#isArchived() archived}.
+     */
+    protected final void markArchived() {
+        currentOperation().markArchived();
+    }
+
+    /**
+     * Marks the entity under migration as {@linkplain Entity#isDeleted() deleted}.
+     */
+    protected final void markDeleted() {
+        currentOperation().markDeleted();
+    }
+
+    /**
+     * Configures the migration to delete the entity record from the storage.
+     *
+     * <p>The entity modifications are still applied, allowing to trigger
+     * {@linkplain EntityLifecycle entity lifecycle} events before the actual record deletion.
+     */
+    protected final void removeFromStorage() {
+        currentOperation().removeFromStorage();
+    }
+
+    /**
+     * Returns the ID of an entity under migration.
+     */
+    protected final I id() {
+        return currentOperation().id();
+    }
+
+    /**
+     * Returns the version of entity under migration.
+     */
+    protected final Version version() {
+        return currentOperation().version();
+    }
+
+    /**
+     * Returns {@code true} if the entity under migration is
+     * {@linkplain Entity#isArchived() archived}.
+     */
+    protected final boolean isArchived() {
+        return currentOperation().isArchived();
+    }
+
+    /**
+     * Returns {@code true} if the entity under migration is
+     * {@linkplain Entity#isDeleted() deleted}.
+     */
+    protected final boolean isDeleted() {
+        return currentOperation().isDeleted();
+    }
+
+    final boolean physicallyRemoveRecord() {
+        return currentOperation().physicallyRemoveRecord();
+    }
+
+    final void finishCurrentOperation() {
```

Comment:
Looks like this method should be called by the caller code to finalize the operation. Am I right? If so, it worth documenting the contract.